### ISSUE TITLE
Report a small bug, and a small fix: False alarm "could not determine the type of this expression"

### ISSUE THREAD

**Issue by gilad-bendor** [google/closure-compiler#1685]
Hello,
I encountered this annoying bug many times - the bug can be demonstrated by compiling this code:

	(function(){
		/** @constructor */
		function MyClass() {}
		/** @type {!string} */ MyClass.prototype.myProp;
		/** @type {!MyClass} */({}).myProp;                            <== Error: "could not determine the type of this expression"
	})();

After long deep effort, it seems that this can be fixed by adding one line to com.google.javascript.jscomp.TypeInference:
Add this line:
	objType.resolve(registry.getErrorReporter(), scope);
right after this line:
	if (propertyType == null && objType != null) {
in the method 
	private JSType getPropertyType(JSType objType, String propName, Node n, FlowScope scope) {...}

**Comment by ChadKillingsworth** [comment]
This is a case of lying to the type system:

```js
/** @type {!MyClass} */({}).myProp;  
```

However, other than the useless code warning, I don't see the issue. Can you give more detail?

**Comment by gilad-bendor** [comment]
body p { margin-bottom: 0cm; margin-top: 0pt; } 
  
  
    Thanks.
    A better example would be:
    
    (function(){
        /** @constructor */
        function MyClass(){}
        /** @type {!string} */ MyClass.prototype.myProperty;
        
        function /** !boolean */ isMyClass(/** * */ x) {
            return (x instanceof MyClass) &&
      (x.myProperty.length > 10);
        }
        
        function myFunction(/** !Object */ myParameter){
            if (isMyClass(myParameter))
                alert(/** @type {!MyClass}
      */(myParameter).myProperty);
        }
        
    })();
    
    
    
    
    
    On 23-Mar-16 14:24, Chad Killingsworth
      wrote:
    
    
      This is a case of lying to the type system:
      
        /** @type {!MyClass} */({}).myProp;  
      
      However, other than the useless code warning, I don't see the
        issue. Can you give more detail?
      —
        You are receiving this because you authored the thread.
        Reply to this email directly or view
          it on GitHub

**Comment by ChadKillingsworth** [comment]
I'm not seeing any issues on the latest version:

http://closure-compiler-debugger.appspot.com/#input0%3D(function()%257B%250A%2520%2520%2520%252F**%2520%2540constructor%2520*%252F%250A%2520%2520%2520function%2520MyClass()%257B%257D%250A%2520%2520%2520%252F**%2520%2540type%2520%257B!string%257D%2520*%252F%2520MyClass.prototype.myProperty%253B%250A%2520%2520%250A%2520%2520%2520function%2520%252F**%2520!boolean%2520*%252F%2520isMyClass(%252F**%2520*%2520*%252F%2520x)%2520%257B%250A%2520%2520%2520%2520%2520%2520%2520return%2520(x%2520instanceof%2520MyClass)%2520%2526%2526%2520(x.myProperty.length%2520%253E%252010)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%2520%2520%2520function%2520myFunction(%252F**%2520!Object%2520*%252F%2520myParameter)%257B%250A%2520%2520%2520%2520%2520%2520%2520if%2520(isMyClass(myParameter))%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520alert(%252F**%2520%2540type%2520%257B!MyClass%257D*%252F(myParameter).myProperty)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%257D)()%253B%250A%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_CONSTANTS%3D1%26CHECK_DEPRECATED%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_GLOBAL_THIS%3D1%26CHECK_MISSING_RETURN%3D1%26CHECK_PROVIDES%3D1%26CHECK_REQUIRES%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_UNREACHABLE_CODE%3D1%26CHECK_VISIBILITY%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1%26TRANSPILE%3D1

Are you perhaps using an old version of the compiler?

**Comment by gilad-bendor** [comment]
body p { margin-bottom: 0cm; margin-top: 0pt; } 
  
  
    It can be seen when you check "ENABLE_ALL_DIAGNOSTIC_GROUPS":
    
http://closure-compiler-debugger.appspot.com/#input0%3D(function()%257B%250A%2520%2520%2520%252F**%2520%2540constructor%2520*%252F%250A%2520%2520%2520function%2520MyClass()%257B%257D%250A%2520%2520%2520%252F**%2520%2540type%2520%257B!string%257D%2520*%252F%2520MyClass.prototype.myProperty%253B%250A%2520%2520%250A%2520%2520%2520function%2520%252F**%2520!boolean%2520*%252F%2520isMyClass(%252F**%2520*%2520*%252F%2520x)%2520%257B%250
A%2520%2520%2520%2520%2520%2520%2520return%2520(%2520x%2520instanceof%2520MyClass)%2520%2526%2526%2520(x.myProperty.length%2520%253E%252010)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%2520%2520%2520function%2520myFunction(%252F**%2520!Object%2520*%252F%2520myParameter)%257B%250A%2520%2520%2520%2520%2520%2520%2520if%2520(isMyClass(myParameter))%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520alert(%252F**%2520%2540type%2520%257B!MyClass%257D*%252F(myParameter).myProperty)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%257D)()%253B%250A%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_CONSTANTS%3D1%26CHECK_DEPRECATED%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_GLOBAL_THIS%3D1%26CHECK_MISSING_RETURN%3D1%26CHECK_PROVIDES%3D1%26CHECK_REQUIRES%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_UNREACHABLE_CODE%3D1%26CHECK_VISIBILITY%3D1%26ENABLE_ALL_DIAGNOSTIC_GROUPS%3D1%26L
ANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26TRANSPILE%3D1
    
    
    On 27-Mar-16 5:55, Chad Killingsworth
      wrote:
    
    
      I'm not seeing any issues on the latest version:
      http://closure-compiler-debugger.appspot.com/#input0%3D(function()%257B%250A%2520%2520%2520%252F**%2520%2540constructor%2520*%252F%250A%2520%2520%2520function%2520MyClass()%257B%257D%250A%2520%2520%2520%252F**%2520%2540type%2520%257B!string%257D%2520*%252F%2520MyClass.prototype.myProperty%253B%250A%2520%2520%250A%2520%2520%2520function%2520%252F**%2520!boolean%2520*%252F%2520isMyClass(%252F**%2520*%2520*%252F%2520x)%2520%
257B%250A%2520%2520%2520%2520%2520%2520%2520return%2520(

          x%2520instanceof%2520MyClass)%2520%2526%2526%2520(x.myProperty.length%2520%253E%252010)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%2520%2520%2520function%2520myFunction(%252F**%2520!Object%2520*%252F%2520myParameter)%257B%250A%2520%2520%2520%2520%2520%2520%2520if%2520(isMyClass(myParameter))%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520alert(%252F**%2520%2540type%2520%257B!MyClass%257D*%252F(myParameter).myProperty)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%257D)()%253B%250A%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_CONSTANTS%3D1%26CHECK_DEPRECATED%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_GLOBAL_THIS%3D1%26CHECK_MISSING_RETURN%3D1%26CHECK_PROVIDES%3D1%26CHECK_REQUIRES%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_UNREACHABLE_CODE%3D1%26CHECK_VISIBILITY%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26
PRETTY_PRI

          NT%3D1%26TRANSPILE%3D1
      Are you perhaps using an old version of the compiler?
      —
        You are receiving this because you authored the thread.
        Reply to this email directly or view
          it on GitHub

**Comment by gilad-bendor** [comment]
body p { margin-bottom: 0cm; margin-top: 0pt; } 
  
  
    Again, it can be seen when you check "ENABLE_ALL_DIAGNOSTIC_GROUPS":
    
    http://closure-compiler-debugger.appspot.com/#input0%3D(function()%257B%250A%2520%2520%2520%252F**%2520%2540constructor%2520*%252F%250A%2520%2520%2520function%2520MyClass()%257B%257D%250A%2520%2520%2520%252F**%2520%2540type%2520%257B!string%257D%2520*%252F%2520MyClass.prototype.myProperty%253B%250A%2520%2520%250A%2520%2520%2520function%2520%252F**%2520!boolean%2520*%252F%2520isMyClass(%252F**%2520*%2520*
%252F%2520x)%2520%257B%250
A%2520%2520%2520%2520%2520%2520%2520return%2520(%2520x%2520instanceof%2520MyClass)%2520%2526%2526%2520(x.myProperty.length%2520%253E%252010)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%2520%2520%2520function%2520myFunction(%252F**%2520!Object%2520*%252F%2520myParameter)%257B%250A%2520%2520%2520%2520%2520%2520%2520if%2520(isMyClass(myParameter))%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520alert(%252F**%2520%2540type%2520%257B!MyClass%257D*%252F(myParameter).myProperty)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%257D)()%253B%250A%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_CONSTANTS%3D1%26CHECK_DEPRECATED%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_GLOBAL_THIS%3D1%26CHECK_MISSING_RETURN%3D1%26CHECK_PROVIDES%3D1%26CHECK_REQUIRES%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_UNREACHABLE_CODE%3D1%26CHECK_VISIBILITY%3D1%26ENABLE_ALL_DIAGNOSTIC_GROUPS%3D1%26
L
ANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26TRANSPILE%3D1
    
    :-)
    
    
    On 27-Mar-16 5:55, Chad Killingsworth
      wrote:
    
    
      I'm not seeing any issues on the latest version:
      http://closure-compiler-debugger.appspot.com/#input0%3D(function()%257B%250A%2520%2520%2520%252F**%2520%2540constructor%2520*%252F%250A%2520%2520%2520function%2520MyClass()%257B%257D%250A%2520%2520%2520%252F**%2520%2540type%2520%257B!string%257D%2520*%252F%2520MyClass.prototype.myProperty%253B%250A%2520%2520%250A%2520%2520%2520function%2520%252F**%2520!boolean%2520*%252F%2520isMyClass(%252F**%2520*%2520*%252F%2520x)%2520%
257B%250A%2520%2520%2520%2520%2520%2520%2520return%2520(

          x%2520instanceof%2520MyClass)%2520%2526%2526%2520(x.myProperty.length%2520%253E%252010)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%2520%2520%2520function%2520myFunction(%252F**%2520!Object%2520*%252F%2520myParameter)%257B%250A%2520%2520%2520%2520%2520%2520%2520if%2520(isMyClass(myParameter))%250A%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520%2520alert(%252F**%2520%2540type%2520%257B!MyClass%257D*%252F(myParameter).myProperty)%253B%250A%2520%2520%2520%257D%250A%2520%2520%250A%257D)()%253B%250A%250A%26input1%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_CONSTANTS%3D1%26CHECK_DEPRECATED%3D1%26CHECK_ES5_STRICT%3D1%26CHECK_GLOBAL_NAMES%3D1%26CHECK_GLOBAL_THIS%3D1%26CHECK_MISSING_RETURN%3D1%26CHECK_PROVIDES%3D1%26CHECK_REQUIRES%3D1%26CHECK_SUSPICIOUS_CODE%3D1%26CHECK_SYMBOLS%3D1%26CHECK_TYPES%3D1%26CHECK_UNREACHABLE_CODE%3D1%26CHECK_VISIBILITY%3D1%26LANG_IN_IS_ES6%3D1%26MISSING_PROPERTIES%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26
PRETTY_PRI

          NT%3D1%26TRANSPILE%3D1
      Are you perhaps using an old version of the compiler?
      —
        You are receiving this because you authored the thread.
        Reply to this email directly or view
          it on GitHub

**Comment by gilad-bendor** [comment]
body p { margin-bottom: 0cm; margin-top: 0pt; } 
  
  
    Lying to the type system is an established and honorable
    institution...
    Google-closure would have become useless if not for that.
    From my experience with Google-Closure, you just can't avoid that
    (at least elegantly).
    Life is not a fairy-tale   ;-)
    
    In my Eclipse environment, I have set up a "builder" that can
    suppress Google-Closure warnings - like this (note the wildcard
    "*"):
            myFunc();        // SUPRESS: variable * is undeclared
    
    This feature (which - in my opinion - is a MUST in Google-Closure)
    is extremely helpful in that it allows me to keep my (very large)
    code without any warnings.
    This way, when a new warning (which is obviously "interesting" to
    me) is created - I can see it immediately.
    The alternative would be hundreds of
    warnings-that-will-never-go-away, and new MEANINGFUL warnings  -
    which just can't be noticed.
    
    I am giving you a feedback that you are used to ignore: you want the
    code to be totally clean, without "lying".
    However - this is how Google-Closure is at its best: shining in the
    real world.
    I would *never* have made it without your wonderful tool,
     however I would never be able to use Google-Closure without
    "lying"...
    
    In my view, the flag "ENABLE_ALL_DIAGNOSTIC_GROUPS" is very VERY
    important, and yields the best warnings.
    However - you chose to un-set it by default.
    And why ?
    The answer is: it generates many warnings which can't (at least not
    elegantly) be fixed.
    I my experience in the real world, this bug that I reported
    eliminates many of these "false alarms" and is very important.
    
    Once you support the "// SUPRESS: ..." feature, one could set the
    "ENABLE_ALL_DIAGNOSTIC_GROUPS" flag without being flooded in
    warnings.
    And once you do that, by bug report would become much more
    significant.
    
    Please don't isolate yourselves in the "pure code" ivory tower: make
    your awesome tool more usable in the real world !
    
    
    
    On 23-Mar-16 14:24, Chad Killingsworth
      wrote:
    
    
      This is a case of lying to the type system:
      
        /** @type {!MyClass} */({}).myProp;  
      
      However, other than the useless code warning, I don't see the
        issue. Can you give more detail?
      —
        You are receiving this because you authored the thread.
        Reply to this email directly or view
          it on GitHub
